src/client_side.cc

```diff
@@ -3050,7 +3057,10 @@ ConnStateData::switchToHttps(HttpRequest *request, Ssl::BumpMode bumpServerMode)
 {
     assert(!switchedToHttps_);
 
-    sslConnectHostOrIp = request->url.host();
+    static char ip[MAX_IPSTRLEN];
+    sslConnectHostOrIp =  request->url.hostIsNumeric() ?
+        request->url.hostIP().toStr(ip, sizeof(ip)) :request->url.host();
```

Comment:
Please avoid duplicating this code -- add something like a `url.hostOrIp()` method if needed. BTW, why do you need to change sslConnectHostOrIp value? Does that change fix a bug?

Reply:
OK I added and use the Anyp::Uri::hostOrIp method.

Also I changed the way the ClientHttpRequest::uri computed inside buildFakeRequest was buggy.

> why do you need to change sslConnectHostOrIp value? Does that change fix a bug?

Nope. just to use it with initiateTunneledRequest.

But I just realised that the sslConnectHostOrIp is used to generate certificates.  This fix affects ssl_crtd daemon results in the case we are connecting to an ipv6 TLS server and no SNI exist nor a hostname in related CONNECT request.

Current squid code in this case will generate '[ipv6_address]' as sslConnectHostOrIp (inside '[]' brackets) and this patch will changes this behaviour and will generate a simple 'ipv6_address'. This is will cause squid to ignore currently cached certificates and generate new ones.

My opinion is that this is not a huge problem and it will happen only for a small part of cached certificates.

Also looks that we are not generating correctly such certificates (with ip as CN which is not checked by browsers) but we have to build also a subjectAltName with an ip entry:
    https://tools.ietf.org/html/rfc5280#section-4.2.1.6

Opinions?

Reply:
> Also looks that we are not generating correctly such certificates (with ip as CN which is not checked by browsers) but we have to build also a subjectAltName with an ip entry:
https://tools.ietf.org/html/rfc5280#section-4.2.1.6

> Opinions?

If I understand the question correctly, you are asking whether Squid should change how IP-based certificates are generated. I think that question should not be asked during this PR. We have enough problems here without that question! I do not know the answer to the question though -- the RFC text you linked to appears to talk about how to add an IP address as an Subject Alternative Name, not which fields/names should be used when generating an IP-based certificate. If you agree that this problem is out of this PR scope, then let's not try to answer that question here. Please file a bug report instead.

Reply:
I am resolving this change request by adding the following paragraph to the PR description (based on my interpretation of your answer). As always, please edit the PR description as needed so that it reflects your understanding of the current PR state.

> Also changed how IPv6-based certificates are generated. Their CN field value is no longer surrounded by [square brackets]. This change was done to streamline Squid code. It affects certificate cache key so old IPv6-based certificates will never be found (and will eventually be purged) while new ones will be generated and cached instead. We believe these IPv6-based certificates are rare and untrusted by browsers so the change in their CN should not have a significant affect on users.